crates/sui-network/src/randomness/mod.rs

```diff
@@ -0,0 +1,686 @@
+                    authority_info,
+                    epoch,
+                    round,
+                    partial_sigs
+                ))
+            });
+        }
+
+        if let Some(last_handled_key) = last_handled_key {
+            // Remove stuff from the pending_tasks map that we've handled.
+            let split_point = self
+                .pending_tasks
+                .range((Bound::Excluded(last_handled_key), Bound::Unbounded))
+                .next()
+                .cloned();
+            if let Some(key) = split_point {
+                self.pending_tasks = self.pending_tasks.split_off(&key);
+            } else {
+                self.pending_tasks.clear();
+            }
+        }
+        self.update_rounds_pending_metric();
+
+        // After starting a round, we have generated our own partial sigs. Check if that's
+        // enough for us to aggregate already.
+        for (epoch, round) in rounds_to_aggregate {
+            self.maybe_aggregate_partial_signatures(epoch, round);
+        }
+    }
+
+    async fn send_partial_signatures_task(
+        name: AuthorityName,
+        network: anemo::Network,
+        metrics: Metrics,
+        authority_info: Arc<HashMap<AuthorityName, (PeerId, PartyId)>>,
+        epoch: EpochId,
+        round: RandomnessRound,
+        partial_sigs: Vec<RandomnessPartialSignature>,
+    ) {
+        let _metrics_guard = metrics
+            .round_observation_latency_metric()
+            .map(|metric| metric.start_timer());
+
+        let peers: HashMap<_, _> = authority_info
+            .iter()
+            .map(|(name, (peer_id, _party_id))| (name, network.waiting_peer(*peer_id)))
+            .collect();
+        let partial_sigs: Vec<_> = partial_sigs
+            .iter()
+            .map(|sig| bcs::to_bytes(sig).expect("message serialization should not fail"))
+            .collect();
+
+        loop {
+            let mut requests = FuturesUnordered::new();
+            for (peer_name, peer) in &peers {
+                if name == **peer_name {
+                    continue; // don't send partial sigs to self
+                }
+                let mut client = RandomnessClient::new(peer.clone());
+                const SEND_PARTIAL_SIGNATURES_TIMEOUT: Duration = Duration::from_secs(10);
+                let request = anemo::Request::new(SendSignaturesRequest {
+                    epoch,
+                    round,
+                    partial_sigs: partial_sigs.clone(),
+                    sig: None,
+                })
+                .with_timeout(SEND_PARTIAL_SIGNATURES_TIMEOUT);
+                requests.push(async move {
+                    let result = client.send_signatures(request).await;
+                    if let Err(e) = result {
+                        debug!("failed to send partial signatures to {peer_name}: {e:?}");
+                    }
+                });
+            }
+
+            while requests.next().await.is_some() {
```

Comment:
i think we should add an explicit top level timeout here instead of using an individual tiemout on each request. it makes it clearer (imo) that this loop cannot get stuck. additionally it would let you detect whether you timed out, in which case you can skip the sleep and retry immediately. as it is right now, if there were a single bad peer i believe it would cut the retry rate for all the good hosts to 15s instead of 5.

also, you can just use a vector of futures and call `timeout(join_all!(futures), duration)`, the FuturesUnordered isn't really necessary.

Reply:
It's better to use the anemo timeout interface because that allows client to communicate the timeout to the server with the request. I don't think we want to retry immediately if there was not a timeout, because that could lead to infinite fast retry loops until the round is complete when all validators already have the partial sigs and we're just waiting on the checkpoint. Changed to use a `Vec` and `join_all` as suggested.